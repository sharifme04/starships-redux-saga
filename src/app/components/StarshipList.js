import React from "react";
import uuid from "uuid";
import PropTypes from "prop-types";

const StarshipList = props => {
  const { starships } = props;
  const Capitalize = str => str.charAt(0).toUpperCase() + str.slice(1);
  return (
    <div className="mt-1">
      {starships.isLoading ? (
        <div className="text-center">
          <div className="spinner-border custom-spin text-info" role="status">
            <span className="sr-only">Loading...</span>
          </div>
        </div>
      ) : (
        <div className="row justify-content-center">
          {starships && starships.starships && starships.starships.results ? (
            starships.starships.results.map(e => (
              <div
                key={uuid.v4()}
                className="col-xs-12 col-sm-12 col-md-6 mt-2"
              >
                <div className="card">
                  <div className="card-body">
                    <h5 className="card-title">Name: {Capitalize(e.name)}</h5>
                    <p className="card-text">
                      Crew: {parseInt(e.crew) ? e.crew : "None"}
                    </p>
                    <p className="card-text">
                      Passengers:{" "}
                      {parseInt(e.passengers) ? e.passengers : "None"}
                    </p>
                    <div style={{ display: "flex" }}>
                      <p className="card-text">Hyperdrive Class:</p>
                      {e.hyperdrive_rating ? (
                        <div className="col-md-9 mt-1">
                          <div className="progress">
                            <div
                              className="progress-bar bg-info"
                              role="progressbar"
                              style={{
                                width: `${(parseFloat(e.hyperdrive_rating) *
                                  100) /
                                  6}%`
                              }}
                              aria-valuenow={
                                (parseFloat(e.hyperdrive_rating) * 100) / 6
                              }
                              aria-valuemin="0"
                              aria-valuemax="100"
                            ></div>
                          </div>
                        </div>
                      ) : (
                        <div className="col-md-9 mt-1">
                          <div className="progress custom-progress">
                            <div
                              className="progress-bar bg-info"
                              role="progressbar"
                              style={{ width: `0%` }}
                              aria-valuenow="0"
                              aria-valuemin="0"
                              aria-valuemax="100"
                            ></div>
                          </div>
                        </div>
                      )}
                    </div>
                  </div>
                </div>
              </div>
            ))
          ) : (
            <h5> Starships are not available</h5>
          )}
        </div>
      )}
    </div>
  );
};

StarshipList.propTypes = {
  starships: PropTypes.object,
  results: PropTypes.array,
  isLoading: PropTypes.bool
};

export default StarshipList;
