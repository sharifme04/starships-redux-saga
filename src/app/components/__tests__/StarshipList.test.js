import React from "react";
import { shallow } from "enzyme";
import StarshipList from "../StarshipList";

jest.mock("uuid", () => {
  return {
    v4: jest.fn(() => 1)
  };
});

const fakeShips = {
  isLoading: false,
  starships: {
    count: 37,
    next: "https://swapi.co/api/starships/?page=2",
    previous: null,
    results: [
      {
        name: "Executor",
        model: "Executor-class star dreadnought",
        manufacturer: "Kuat Drive Yards, Fondor Shipyards",
        cost_in_credits: "1143350000",
        length: "19000",
        max_atmosphering_speed: "n/a",
        crew: "279144",
        passengers: "38000",
        cargo_capacity: "250000000",
        consumables: "6 years",
        hyperdrive_rating: "2.0",
        MGLT: "40",
        starship_class: "Star dreadnought",
        pilots: [],
        films: [
          "https://swapi.co/api/films/2/",
          "https://swapi.co/api/films/3/"
        ],
        created: "2014-12-15T12:31:42.547000Z",
        edited: "2017-04-19T10:56:06.685592Z",
        url: "https://swapi.co/api/starships/15/"
      }
    ]
  }
};

describe("<StarshipList />", () => {
  test("renders", () => {
    const component = shallow(<StarshipList starships={fakeShips} />);
    expect(component).toMatchSnapshot();
  });

  test("return the default array when there is no data to map through", () => {
    const component = shallow(<StarshipList />);
    expect(component).toMatchSnapshot();
  });

  test("does not break without value", () => {
    const component = shallow(<StarshipList />);
    expect(component.find("h5")).toHaveLength(1);
    expect(component.find("p")).toHaveLength(0);
  });

  test("does not break with an empty array", () => {
    const component = shallow(<StarshipList results={[]} />);
    expect(component.find("h5")).toHaveLength(1);
    expect(component.find("p")).toHaveLength(0);
  });
});
