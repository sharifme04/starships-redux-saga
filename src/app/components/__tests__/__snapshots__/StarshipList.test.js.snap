// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StarshipList /> renders 1`] = `
<div
  className="mt-1"
>
  <div
    className="row justify-content-center"
  >
    <div
      className="col-xs-12 col-sm-12 col-md-6 mt-2"
      key="1"
    >
      <div
        className="card"
      >
        <div
          className="card-body"
        >
          <h5
            className="card-title"
          >
            Name: 
            Executor
          </h5>
          <p
            className="card-text"
          >
            Crew: 
            279144
          </p>
          <p
            className="card-text"
          >
            Passengers:
             
            38000
          </p>
          <div
            style={
              Object {
                "display": "flex",
              }
            }
          >
            <p
              className="card-text"
            >
              Hyperdrive Class:
            </p>
            <div
              className="col-md-9 mt-1"
            >
              <div
                className="progress"
              >
                <div
                  aria-valuemax="100"
                  aria-valuemin="0"
                  aria-valuenow={33.333333333333336}
                  className="progress-bar bg-info"
                  role="progressbar"
                  style={
                    Object {
                      "width": "33.333333333333336%",
                    }
                  }
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<StarshipList /> return the default array when there is no data to map through 1`] = `
<div
  className="mt-1"
>
  <div
    className="row justify-content-center"
  >
    <h5>
       Starships are not available
    </h5>
  </div>
</div>
`;
